File: jonghyun/11725 트리의 부모 찾기.py
from collections import deque

class Node:
    def __init__(self):
        self.parent=0
        self.link=[]
        
n=int(input())

graph=[Node() for _ in range(n)]

for _ in range(n-1):
    node1,node2=map(int, input().split(" "))
    graph[node1-1].link.append(node2-1)     ## @ 1번 노드의 인덱스는 0이 되도록 조정 - 인덱스 조정은 입력, 출력 부분에서만
    graph[node2-1].link.append(node1-1)
    
    
    
q=deque()

##최초의 노드 인덱스를 큐에 넣기
idx_now=0
q.append(idx_now)

while len(q):
    idx_now=q.popleft()
    
    for i in graph[idx_now].link:
        if i == graph[idx_now].parent:      ##현재 노드와 연결되어있는 노드들 중 부모인 노드를 제외한다
            continue
        
        q.append(i)
        graph[i].parent=idx_now
        
#DEBUG
debug_count=1
for i in graph:
    print("num: %d, parent: %d, link: "%(debug_count, i.parent+1) + str(i.link))
    debug_count+=1
        
flag_first_skipped=False
for i in graph:
    if flag_first_skipped:
        print(str(i.parent+1))      ## @ 조정한 인덱스 복귀
    else:
        flag_first_skipped=True

File: jonghyun/11725 트리의 부모 찾기.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>

using namespace std;

struct Node {
	int parent;
	vector<int> link;
};

int main() {
	int n;
	cin >> n;

	vector<Node> graph(n);
	int node1, node2;
	for (int i = 0; i < n-1; i++) {
		cin >> node1 >> node2;
		node1--;		//1번 노드는 인덱스가 0이 되기 때문에
		node2--;

		graph[node1].link.push_back(node2);
		graph[node2].link.push_back(node1);
	}

	queue<int> q;

	//최초의 값 큐에 넣기 - 여기서는 인덱스 0번 노드가 처음이다
	int idx_now = 0;
	q.push(idx_now);

	int debug_count = 0;
	//BFS 실행
	while (q.size()) {
		idx_now = q.front();

		//cout << "[" << debug_count++ << "] idx: " << idx_now << ", parent: " << graph[idx_now].parent+1 << "\n";

		q.pop();

		for (auto& i : graph[idx_now].link) {
			if (i == graph[idx_now].parent) {
				continue;
			}

			q.push(i);
			graph[i].parent = idx_now;
		}		
	}

	bool flag_skipped_first = false;
	for (auto& i : graph) {
		if (flag_skipped_first) {
			cout << i.parent + 1 << "\n";
		}
		else {
			flag_skipped_first = true;
		}
	}

	return 0;
}